src/lib.rs

```diff
@@ -27,19 +27,41 @@ use failure::{bail, Fail, Fallible};
 use rayon::prelude::*;
 use std::{
     fmt, mem,
+    num::NonZeroUsize,
     ops::{Index, IndexMut},
 };
 
 #[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone, Debug, Hash)]
 #[cfg_attr(feature = "deser", derive(Deserialize, Serialize))]
 /// A node identifier within a particular `Arena`
 pub struct NodeId {
-    index: usize,
+    /// One-based index.
+    index1: NonZeroUsize,
+}
+
+impl NodeId {
+    /// Returns zero-based index.
+    fn index0(self) -> usize {
+        // This is totally safe because `self.index1 >= 1` is guaranteed by
+        // `NonZeroUsize` type.
+        self.index1.get() - 1
+    }
+
+    /// Creates a new `NodeId` from the given zero-based index.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the given value is `usize::max_value()`.
+    fn from_index0(index0: usize) -> Self {
+        let index1 = NonZeroUsize::new(index0.wrapping_add(1))
+            .expect("Array index is too large");
```

Comment:
We could avoid the panic by returning a `Fallible<NodeId>` from the `new_node` function of `Arena`. This is also a breaking change and I'm not sure if its worth. WDYT?

Reply:
This is theoretically fallible, but I think this hardly happen, because users usually does neither create `usize::max_value()` nodes, nor create `NodeId` for it.
So I don't think it is worth avoiding this panic.

## Another option
If `NodeId::new(index1: NonZeroUsize) -> NodeId` is available, the constructor won't fallible, but then possible panic point moves to `NodeId::new_node()`.

When user added `usize::max_value()`-th node, then `new_node()` would panic.
But I don't think this panic may not happen in real world (we will run out of memory before completing it).

Reply:
Your option sounds very reasonable, let’s go for that.